src/api/mod.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod error;
+mod extractors;
+mod responses;
+mod routes;
+
+use std::{
+    task::Poll,
+    time::SystemTime,
+};
+
+use actix::{
+    Actor,
+    ActorContext,
+    Context,
+    Handler,
+    Message,
+};
+use axum::{
+    http::{
+        Request,
+        Response,
+    },
+    Server,
+};
+use error::ListenerError;
+use futures::Future;
+use hyper::{
+    Method,
+    Uri,
+};
+use lazy_static::lazy_static;
+use mongodb::{
+    options::ClientOptions,
+    Client,
+};
+use pin_project::pin_project;
+use prometheus::{
+    Gauge,
+    HistogramOpts,
+    HistogramVec,
+    IntCounter,
+    IntCounterVec,
+    Opts,
+    Registry,
+};
+use routes::routes;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use tokio::sync::oneshot;
+use tower::{
+    Layer,
+    Service,
+};
+
+use crate::config::mongo::MongoConfig;
+
+/// The Chronicle API actor
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct ChronicleAPI {
+    mongo_config: MongoConfig,
+    #[serde(skip)]
+    server_handle: Option<oneshot::Sender<()>>,
+}
+
+impl ChronicleAPI {
+    pub fn new(mongo_config: MongoConfig) -> Self {
+        Self {
+            mongo_config,
+            server_handle: None,
+        }
+    }
+}
+
+impl PartialEq for ChronicleAPI {
+    fn eq(&self, other: &Self) -> bool {
+        self.mongo_config == other.mongo_config
+    }
+}
+
+impl Clone for ChronicleAPI {
+    fn clone(&self) -> Self {
+        Self {
+            mongo_config: self.mongo_config.clone(),
+            server_handle: None,
+        }
+    }
+}
+
+impl Actor for ChronicleAPI {
+    type Context = Context<Self>;
+
+    fn started(&mut self, _ctx: &mut Self::Context) {
+        register_metrics();
+        let client_opts: ClientOptions = self.mongo_config.clone().into();
+        log::info!("Connecting to MongoDB");
+        let client = Client::with_options(client_opts)
+            .map_err(|e| {
+                log::error!("Unable to create client from options, error: {}", e);
+            })
+            .unwrap();
```

Comment:
Move these over to the construction of the `ChronicleAPI` so that we can avoid the `.unwrap()` calls. For example like: `ChronicleAPI::new() -> Result<_,_>`.